### PR TITLE
HTTPS client support

### PR DESCRIPTION
**Pull request by jukkar**
This PR adds HTTPS support to http library. It also creates a https_client sample application.

### CODE REVIEW

File: include/net/http.h [hunk 1]
```diff
@@ -418,6 +526,45 @@ static inline int http_client_send_post_req(struct http_client_ctx *http_ctx,
...
 		     const char *server, u16_t server_port);
 
+#if defined(CONFIG_HTTPS)
+/**
+ * @brief Initialize user supplied HTTP context when using HTTPS.
+ *
+ * @detail Caller can set the various fields in http_ctx after this call
+ * if needed.
+ *
+ * @param http_ctx HTTPS context.
```
Comment (pfalcon) [inline include/net/http.h]: In the prototype below this param is called just "ctx".

File: include/net/http.h [hunk 2]
```diff
@@ -277,13 +382,16 @@ struct http_client_request {
...
  *
+ * @param ctx HTTP client context.
  * @param net_ctx Network context.
  * @param req HTTP request to perform.
  * @param timeout Timeout when doing net_buf allocations.
  *
  * @return Return 0 if ok, and <0 if error.
  */
-int http_request(struct net_context *net_ctx, struct http_client_request *req,
+int http_request(struct http_client_ctx *ctx,
```
Comment (pfalcon) [inline include/net/http.h]: Would it make sense to make net_contex* be part of http_client_ctx and not bother user to pass both?

**Review by pfalcon — APPROVED** [review]
Eyeballed. This all looks complicated, I don't know why need to wrap mbedTLS behind such facade API (a function which takes 10+ arguments), but assume that's what's needed.

Few specific suggestions in the inline comments.

File: samples/net/https_client/README.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+Open the :file:`src/config.h` file and set the server port
+to match the HTTP server setup, for example:
+
+.. code-block:: c
+
+   #define SERVER_PORT		443
+
+assumes that the HTTPS server is listening at the TCP port 443.
+If the default example HTTPS server is used, then the default
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: better would be to add "(described in the next section)"

File: samples/net/https_client/README.rst [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+port is 4443.
+
+HTTPS Server
+============
+
+A very simple HTTPS server is provided in net-tool project.
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: Combine this into:

Sample code for a very simple HTTPS server can be downloaded from the zephyrproject-rtos/net-tools project area: https://github.com/zephyrproject-rtos/net-tools

File: samples/net/https_client/README.rst [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+The net-tools can be downloaded from
+
+    https://github.com/zephyrproject-rtos/net-tools
+
+
+Open a terminal window and type:
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: Isn't there some configuration you need to do before just starting up the HTTPS server?

File: samples/net/https_client/README.rst [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+.. code-block:: console
+
+   $ cd net-tools
+   $ ./https-server.sh
+
+
+DNS setup
+=========
+
+The net-tool project provides a simple DNS resolver. You can activate
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: it's "net-tools" (plural), yes?

File: samples/net/https_client/README.rst [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+it like this if you want to test the DNS resolving with HTTP client.
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: ... to test DNS resolving with the HTTP client.

**Review by dbkinder — CHANGES_REQUESTED** [review]
a few suggestions

File: samples/net/https_client/README.rst (hunk 5 above)
Comment (jukkar) [inline samples/net/https_client/README.rst]: Indeed there is the slip setup that is not described in this file. I will add that info here too.
I also noticed that slip setup info is also missing from http-client, and http-server samples README.rst file, I will send a separate patch for that.

**Comment by jukkar** [comment]
> Eyeballed. This all looks complicated, I don't know why need to wrap mbedTLS behind such facade API (a function which takes 10+ arguments), but assume that's what's needed.

Setting up this beast is complicated, what is now in https_client_init() is quite minimal params to have. I tried to avoid introducing too much mbedtls specific calls in applications. Unfortunately the https_client_init() would have been even more convoluted if the certification setup would have been added to the API. That is why there is a separate cert callback which application must have which can do various cert things (if needed).

File: samples/net/https_client/README.rst [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+to an HTTPS server and how to parse the incoming responses.
+Supported HTTP 1.1 methods are: GET, HEAD, OPTIONS and POST.
+
+The source code for this sample application can be found at:
+:file:`samples/net/https_client`.
+
+Requirements
+************
+
+- :ref:`Networking with Qemu <networking_with_qemu>`
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: PR 415 is making the same change?   Same comment -- Qemu should be spelled QEMU.  If you just use :ref:`networking_with_qemu` the output will be the title of that section (which is "Networking with QEMU")

File: samples/net/https_client/README.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+- HTTPS Server
+- DNS server (optional)
+
+Building and Running
+********************
+
+Open the project configuration file for your platform, for example:
+:file:`prj_qemu_x86.conf` is the configuration file for QEMU.
+
+To use QEMU for testing, follow the :ref:`Networking with Qemu
```
Comment (dbkinder) [inline samples/net/https_client/README.rst]: Qemu -> QEMU

**Review by dbkinder — CHANGES_REQUESTED** [review]
PR 415 is also making this change?

File: samples/net/https_client/README.rst (hunk 9 above)
Comment (jukkar) [inline samples/net/https_client/README.rst]: Ouch, I copied the ref from some old sample which obviously was wrong already.
I will fix all invocations of those incorrectly typed refs in PR #415

File: samples/net/https_client/README.rst (hunk 8 above)
Comment (jukkar) [inline samples/net/https_client/README.rst]: @dbkinder Thanks for the hint, I am no expert on these rst formats. I will remove the extra title from the ref then.

**Comment by jukkar** [comment]
I uploaded new version that should address all the issues people commented.

**Review by pfalcon — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Pull request merged**